### PR TITLE
Add --skipLibCheck compiler option

### LINKED ISSUE #8521: Slow compilation with really huge declaration files
Hello,

The NativeScript framework exposes all native Android and iOS APIs to JavaScript. To take use of the TypeScript features we have tools to generate TypeScript declaration files. These declaration files are verified to have no syntax or type errors when they are generated.

Compiling a small `ts` file that references types from the big `d.ts`-es results in about 15 sec. compilation time on our side. Anything that can help us to speed up the build time would be greatly appreciated.

This is how the diagnostics of a project using our `.d.ts`-es looks like:
```
nativescript-big-dts cankov$ tsc
Files:               4
Lines:          317792
Nodes:         1235680
Identifiers:    486912
Symbols:        709471
Types:          145659
Memory used:   824485K
I/O read:        0.02s
I/O write:       0.00s
Parse time:      2.04s
Bind time:       1.32s
Check time:     12.09s
Emit time:       0.03s
Total time:     15.49s
```

I have prepared a small project suitable for profiling here:
https://github.com/PanayotCankov/nativescript-big-dts

Thanks,

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR adds a new `--skipLibCheck` compiler option that causes type checking of declaration files (files with extension `.d.ts`) to be skipped. When a program includes large declaration files, the compiler spends a lot of time type checking declarations that are already known to not contain errors, and a compile times may be significantly shortened by skipping declaration file type checks.

Since declarations in one file can affect type checking in other files, some errors may not be detected when `--skipLibCheck` is specified. For example, if a non-declaration file augments a type declared in a declaration file, errors may result that are only reported when the declaration file is checked. However, in practice such situations are rare.

Fixes #8521. Compile times for the example given in this issue are reduced 5x.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
👍

**Comment by mhegazy** [comment]
:+1:

**Comment by vladima** [comment]
:+1:

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16383,14 +16383,11 @@ namespace ts {
...
-                // and whether the user has specifically chosen to avoid checking it.
-                if (compilerOptions.skipDefaultLibCheck) {
-                    // If the user specified '--noLib' and a file has a '/// <reference no-default-lib="true"/>',
-                    // then we should treat that file as a default lib.
-                    if (node.hasNoDefaultLib) {
-                        return;
-                    }
+                // If skipLibCheck is enabled, skip type checking if file is a declaration file.
+                // If skipDefaultLibCheck is enabled, skip type checking if file contains a
+                // '/// <reference no-default-lib="true"/>' directive.
```
Comment (basarat) [inline src/compiler/checker.ts]: Curious why we would skip for `noDefaultLib`.  Is this to support just test harnesses or actual user code? :rose:

**Pull request closed without merging**